pkg/antctl/raw/featuregates/command.go

```diff
@@ -0,0 +1,207 @@
+
+func init() {
+	Command = &cobra.Command{
+		Use:   "featuregates",
+		Short: "Get feature gates list",
+	}
+
+	if runtime.Mode == runtime.ModeAgent {
+		Command.RunE = agentRunE
+		Command.Long = "Get current Antrea agent feature gates info"
+	} else if runtime.Mode == runtime.ModeController && runtime.InPod {
+		Command.RunE = controllerLocalRunE
+		Command.Long = "Get current Antrea controller feature gates info."
+	} else if runtime.Mode == runtime.ModeController && !runtime.InPod {
+		Command.Long = "Get Antrea feature gates info including Controller and Agent"
+		Command.RunE = controllerRemoteRunE
+	}
+}
+
+func setupKubeconfig(kubeconfig *rest.Config) {
+	kubeconfig.Insecure = true
+	kubeconfig.CAFile = ""
+	kubeconfig.CAData = nil
+	kubeconfig.GroupVersion = &schema.GroupVersion{Group: "", Version: ""}
+	kubeconfig.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	if runtime.InPod {
+		if runtime.Mode == runtime.ModeAgent {
+			kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaAgentAPIPort))
+			kubeconfig.BearerTokenFile = agentapiserver.TokenPath
+		} else {
+			kubeconfig.Host = net.JoinHostPort("127.0.0.1", strconv.Itoa(apis.AntreaControllerAPIPort))
+			kubeconfig.BearerTokenFile = controllerapiserver.TokenPath
+		}
+	}
+}
+
+func agentRunE(cmd *cobra.Command, _ []string) error {
+	return featureGateRequest(cmd, runtime.ModeAgent)
+}
+
+func controllerLocalRunE(cmd *cobra.Command, _ []string) error {
+	return featureGateRequest(cmd, runtime.ModeController)
+}
+
+func controllerRemoteRunE(cmd *cobra.Command, _ []string) error {
+	return featureGateRequest(cmd, "remote")
+}
+
+func featureGateRequest(cmd *cobra.Command, mode string) error {
+	kubeconfigPath, err := cmd.Flags().GetString("kubeconfig")
+	if err != nil {
+		return err
+	}
+	kubeconfig, err := runtime.ResolveKubeconfig(kubeconfigPath)
+	if err != nil {
+		return err
+	}
+	restconfigTmpl := rest.CopyConfig(kubeconfig)
+	setupKubeconfig(restconfigTmpl)
+	if server, err := Command.Flags().GetString("server"); err != nil {
+		kubeconfig.Host = server
+	}
+
+	k8sClientset, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to create K8s clientset: %w", err)
+	}
+	antreaClientset, err := antrea.NewForConfig(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("error when creating Antrea clientset: %w", err)
+	}
+
+	var agentResp []featuregates.Response
+	var controllerResp []featuregates.Response
+	var agentClient *rest.RESTClient
+	var controllerClient *rest.RESTClient
+
+	switch mode {
+	case runtime.ModeAgent:
+		agentClient, err = rest.RESTClientFor(restconfigTmpl)
+	case runtime.ModeController:
+		controllerClient, err = rest.RESTClientFor(restconfigTmpl)
+
+	}
+	if err != nil {
+		return fmt.Errorf("fail to creating %s rest client %w", mode, err)
+	}
+
+	if mode == "remote" {
+		if agentClient, err = getAgentClient(k8sClientset, antreaClientset, restconfigTmpl); err != nil {
+			return err
+		}
```

Comment:
I am not sure about this part. It seems that in `getAgentClient` you look for a "random" Node and you select the Agent instance running on that specific Node. I don't think this is great, as the feature gates may not be the same for all agent, if we are in the middle of an upgrade for example. IMO, the best way for "remote" mode would be to only provide the feature gates for the Controller by default, but potentially introduce a new command-line option which would accept a Node name. We would then select the Agent instance running on that specific Node.

Reply:
ah, OK, I thought all agent check the same configmap, so the featuregates should be the same, didn't consider the upgrade case. but it will be more complicate to add a node flag.   looks like @jianjuns 's suggestion https://github.com/vmware-tanzu/antrea/pull/2082#pullrequestreview-638194286 is more applicable to simplify the case, what's your thought?